src/collapse/vsprof.rs

```diff
@@ -0,0 +1,222 @@
+                Ordering::Less => {
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+                // Case 2
+                Ordering::Equal => {
+                    self.write_stack(occurences);
+                    self.stack.pop();
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+                // Case 3
+                Ordering::Greater => {
+                    // The Visual Studio profiler outputs the number of times a function is called.
+                    //
+                    // Let's say we have a function `A()` which is called 500 times, and which
+                    // calls a function `B()` 300 times. If we didn't do anything special here,
+                    // this would result in `A()` being assigned 800 samples, giving the impression
+                    // that `A()` only calls `B()` less than 50% of the time, while in fact it is
+                    // called more than 50% of the time.
+                    //
+                    // To prevent this from happening, we instead subtract the number of calls from
+                    // the previous node (in this case `B()`) from the current node (in this case
+                    // `A()`. This leaves `A()` with the correct number of 500 samples.
+                    //
+                    // If the previous number of calls is equal to the current number of calls, we
+                    // don't want to write the current top node, because that would duplicate the
+                    // number of samples for the current node.
+                    let mut prev_number_of_calls = 0;
+                    for _ in 0..(prev_depth - depth + 1) {
+                        if prev_number_of_calls != self.stack.last().unwrap().1 {
+                            self.write_stack(occurences);
+                        }
+                        prev_number_of_calls = self.stack.pop().unwrap().1;
+
+                        if self.stack.is_empty() {
+                            break;
+                        }
+
+                        let last = self.stack.len() - 1;
+                        let number_of_calls = &self.stack[last].1;
+                        if prev_number_of_calls < *number_of_calls {
+                            self.stack[last].1 -= prev_number_of_calls;
+                        }
+                    }
+
+                    self.stack
+                        .push((function_name.to_string(), number_of_calls));
+                }
+            }
+        } else {
+            return invalid_data_error!("Unable to parse function name from line:\n{}", line);
+        }
+
+        Ok(())
+    }
+
+    // Store the current stack in `occurences`
+    fn write_stack(&self, occurrences: &mut Occurrences) {
+        if let Some(nsamples) = self.stack.last().map(|(_, n)| *n) {
+            if nsamples > 0 {
```

Comment:
If you want you can simplify this somewhat:
```suggestion
        if let Some(nsamples) = self.stack.last().map(|(_, n)| *n).filter(|n| n > 0) {
```